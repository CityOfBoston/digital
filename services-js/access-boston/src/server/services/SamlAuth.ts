import fs from 'fs';
import { DOMParser } from 'xmldom';
import xpath from 'xpath';
import { IdentityProvider, ServiceProvider } from 'saml2-js';

export interface SamlConfigPaths {
  serviceProviderKeyPath: string;
  serviceProviderCertPath: string;
  metadataPath: string;
}

interface SamlResponseHeader {
  version: '2.0';
  destination: string;
  in_response_to: string;
  id: string;
}

interface SamlAuthAssertion {
  response_header: SamlResponseHeader;
  type: 'authn_response';
  user: {
    name_id: string;
    session_index: string;
    attributes: {
      groups?: string[];
    };
  };
}

interface SamlLogoutRequestAssertion {
  response_header: SamlResponseHeader;
  type: 'logout_request';
  issuer: string;
  name_id: string;
  session_index: string;
}

type SamlAssertion = SamlAuthAssertion | SamlLogoutRequestAssertion;

export interface ServiceProviderConfig {
  metadataUrl: string;
  assertUrl: string;
}

export interface SamlLoginResult {
  type: 'login';
  nameId: string;
  sessionIndex: string;
  groups: string[];
}

export interface SamlLogoutRequestResult {
  type: 'logout';
  nameId: string;
  sessionIndex: string;
  successUrl: string;
}

export type SamlAssertResult = SamlLoginResult | SamlLogoutRequestResult;

const SAML_METADATA_NAMESPACES = {
  md: 'urn:oasis:names:tc:SAML:2.0:metadata',
  ds: 'http://www.w3.org/2000/09/xmldsig#',
};

const REDIRECT_BINDING_URI =
  'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect';

export async function makeIdentityProvider(
  metadata: Buffer,
  logoutUrl: string | null = null
): Promise<IdentityProvider> {
  const parser = new DOMParser();
  const select = xpath.useNamespaces(SAML_METADATA_NAMESPACES);
  const doc = parser.parseFromString(metadata.toString('utf-8'));

  const ssoDescriptorElement = select(
    '//md:IDPSSODescriptor',
    doc,
    true
  ) as Element;

  const signRequests =
    ssoDescriptorElement.getAttribute('WantAuthnRequestsSigned') === 'true';

  const redirectBindingElement = select(
    `//md:SingleSignOnService[@Binding='${REDIRECT_BINDING_URI}']`,
    doc,
    true
  ) as Element;

  const redirectUrl = redirectBindingElement.getAttribute('Location') || '';

  const signingCertificateElements = select(
    "//md:KeyDescriptor[@use='signing']",
    doc
  ) as Element[];

  // The textContent is a Base64-encoded binary cert in DER format. We can
  // turn that into PEM just by adding the BEGIN/END bits.
  //
  // If we format the XML then there are some newlines around the Base64
  // content than need to be trimmed.
  const certificates = signingCertificateElements.map(el =>
    [
      '-----BEGIN CERTIFICATE-----',
      el.textContent!.trim(),
      '-----END CERTIFICATE-----',
    ].join('\n')
  );

  return new IdentityProvider({
    sso_login_url: redirectUrl,
    sso_logout_url: logoutUrl || redirectUrl,
    sign_get_request: signRequests,
    certificates,
  });
}

export async function makeServiceProvider(
  { metadataUrl, assertUrl }: ServiceProviderConfig,
  serviceProviderCert: Buffer,
  serviceProviderKey: Buffer
) {
  const privateKey = serviceProviderKey.toString('utf-8');
  const cert = serviceProviderCert.toString('utf-8');

  return new ServiceProvider({
    entity_id: metadataUrl,
    private_key: privateKey,
    certificate: cert,
    assert_endpoint: assertUrl,
    allow_unencrypted_assertion: true,
  });
}

export async function makeSamlAuth(
  {
    serviceProviderKeyPath,
    serviceProviderCertPath,
    metadataPath,
  }: SamlConfigPaths,
  serviceProviderConfig: ServiceProviderConfig,
  logoutUrl: string
): Promise<SamlAuth> {
  const metadata: Promise<Buffer> = new Promise((resolve, reject) => {
    fs.readFile(metadataPath, (err, buf) => {
      if (err) {
        reject(err);
      } else {
        resolve(buf);
      }
    });
  });

  const serviceProviderKey: Promise<Buffer> = new Promise((resolve, reject) => {
    fs.readFile(serviceProviderKeyPath, (err, buf) => {
      if (err) {
        reject(err);
      } else {
        resolve(buf);
      }
    });
  });

  const serviceProviderCert: Promise<Buffer> = new Promise(
    (resolve, reject) => {
      fs.readFile(serviceProviderCertPath, (err, buf) => {
        if (err) {
          reject(err);
        } else {
          resolve(buf);
        }
      });
    }
  );

  const [identityProvider, serviceProvider] = await Promise.all([
    makeIdentityProvider(await metadata, logoutUrl),
    makeServiceProvider(
      serviceProviderConfig,
      await serviceProviderCert,
      await serviceProviderKey
    ),
  ]);

  return new SamlAuth(identityProvider, serviceProvider);
}

export default class SamlAuth {
  private identityProvider: IdentityProvider;
  private serviceProvider: ServiceProvider;

  constructor(
    identityProvider: IdentityProvider,
    serviceProvider: ServiceProvider
  ) {
    this.identityProvider = identityProvider;
    this.serviceProvider = serviceProvider;
  }

  getMetadata(): string {
    return this.serviceProvider.create_metadata();
  }

  makeLoginUrl(): Promise<string> {
    return new Promise((resolve, reject) => {
      this.serviceProvider.create_login_request_url(
        this.identityProvider,
        {},
        (err, loginUrl) => {
          if (err) {
            return reject(err);
          }

          resolve(loginUrl);
        }
      );
    });
  }

  makeLogoutUrl(nameId: string, sessionIndex: string): Promise<string> {
    return new Promise((resolve, reject) => {
      this.serviceProvider.create_logout_request_url(
        this.identityProvider,
        {
          name_id: nameId,
          session_index: sessionIndex,
          sign_get_request: true,
        },
        (err, logoutUrl) => {
          if (err) {
            reject(err);
          } else {
            resolve(logoutUrl);
          }
        }
      );
    });
  }

  protected makeLogoutSuccessUrl(requestId: string): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      this.serviceProvider.create_logout_response_url(
        this.identityProvider,
        {
          in_response_to: requestId,
          sign_get_request: true,
        },
        (err, successUrl) => {
          if (err) {
            reject(err);
          } else {
            resolve(successUrl);
          }
        }
      );
    });
  }

  protected async processSamlAssertion(
    saml: SamlAssertion
  ): Promise<SamlAssertResult> {
    // eslint-disable-next-line no-console
    console.debug('SAML RESPONSE', JSON.stringify(saml, null, 2));

    switch (saml.type) {
      case 'authn_response':
        return {
          type: 'login',
          nameId: saml.user.name_id,
          sessionIndex: saml.user.session_index,
          groups: parseGroupsAttribute(saml.user.attributes.groups || []),
        };
      case 'logout_request':
        return {
          type: 'logout',
          nameId: saml.name_id,
          sessionIndex: saml.session_index,
          successUrl: await this.makeLogoutSuccessUrl(saml.response_header.id),
        };

      default:
        throw new Error(
          `Unrecognized SAML assertion type: ${(saml as any).type}`
        );
    }
  }

  handlePostAssert(body: string): Promise<SamlAssertResult> {
    return new Promise((resolve, reject) => {
      this.serviceProvider.post_assert(
        this.identityProvider,
        { request_body: body },
        (err, saml: SamlAssertion) => {
          if (err) {
            reject(err);
            return;
          }

          try {
            resolve(this.processSamlAssertion(saml));
          } catch (e) {
            reject(e);
          }
        }
      );
    });
  }

  handleGetAssert(query: {
    [key: string]: string | string[];
  }): Promise<SamlAssertResult> {
    return new Promise((resolve, reject) => {
      this.serviceProvider.redirect_assert(
        this.identityProvider,
        { request_body: query },
        (err, saml: SamlAssertion) => {
          if (err) {
            reject(err);
            return;
          }

          try {
            resolve(this.processSamlAssertion(saml));
          } catch (e) {
            reject(e);
          }
        }
      );
    });
  }
}

export function parseGroupsAttribute(groupsAttribute: string[]): string[] {
  // attr here is something like
  // "cn=SG_AB_SERVICEDESK_USERS,cn=Groups,dc=boston,dc=cob"
  return groupsAttribute
    .map(attr => {
      const firstGroupPiece = attr.split(',')[0];
      if (!firstGroupPiece) {
        return '';
      }

      const commonNameMatch = firstGroupPiece.match(/cn=(.*)/);
      return (commonNameMatch && commonNameMatch[1]) || '';
    })
    .filter(s => !!s);
}
