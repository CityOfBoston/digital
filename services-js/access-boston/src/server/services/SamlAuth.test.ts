import { makeIdentityProvider, parseGroupsAttribute } from './SamlAuth';
import path from 'path';
import fs from 'fs';
import crypto from 'crypto';

describe('makeIdentityProvider', () => {
  let samlMetadata: Buffer;

  beforeEach(() => {
    samlMetadata = fs.readFileSync(
      path.join(__dirname, '../../../fixtures/saml-metadata.xml')
    );
  });

  it('parses the xml', async () => {
    const identityProvider = await makeIdentityProvider(samlMetadata);
    expect(identityProvider).toMatchSnapshot();
  });

  it('returns valid certificates', async () => {
    const identityProvider = await makeIdentityProvider(samlMetadata);
    const certificate = identityProvider.certificates[0];

    // We use publicEncrypt just to test that the certificate is parsable.
    const out = crypto.publicEncrypt(certificate, new Buffer('TEST TEXT'));
    expect(out.toString('base64').length).toBeGreaterThan(1);
  });
});

describe('parseGroupsAttribute', () => {
  it('parses out group names', () => {
    const groupsAttribute = [
      'cn=COB-Group-TestGrp01,cn=Groups,dc=boston,dc=cob',
      'cn=SG_AB_IAM_TEAM,cn=Groups,dc=boston,dc=cob',
      'cn=SG_AB_SERVICEDESK_USERS,cn=Groups,dc=boston,dc=cob',
    ];

    expect(parseGroupsAttribute(groupsAttribute)).toEqual([
      'COB-Group-TestGrp01',
      'SG_AB_IAM_TEAM',
      'SG_AB_SERVICEDESK_USERS',
    ]);
  });
});
