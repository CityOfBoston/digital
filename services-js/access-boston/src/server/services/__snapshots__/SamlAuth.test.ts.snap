// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeIdentityProvider parses the xml 1`] = `
IdentityProvider {
  "certificates": Array [
    "-----BEGIN CERTIFICATE-----
MIIHuzCCBaOgAwIBAgITFgAAAO/Occ90ikDr+wAAAAAA7zANBgkqhkiG9w0BAQsFADBwMRMwEQYKCZImiZPyLGQBGRYDY29iMRYwFAYKCZImiZPyLGQBGRYGYm9zdG9uMRgwFgYKCZImiZPyLGQBGRYIY2l0eWhhbGwxJzAlBgNVBAMTHkNpdHlPZkJvc3Rvbi1FbnRlcnByaXNlLVN1Yi1DQTAeFw0xODAyMjYxOTAxMDFaFw0yMTAyMjUxOTAxMDFaMIGAMQswCQYDVQQGEwJVUzELMAkGA1UECBMCTUExDzANBgNVBAcTBkJvc3RvbjEXMBUGA1UEChMOQ2l0eSBPZiBCb3N0b24xEDAOBgNVBAsTB0lBTSBTU08xKDAmBgNVBAMTH3pkcGluZ2ZlZDAxLmNpdHloYWxsLmJvc3Rvbi5jb2IwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQChXwANkULq6nkAZNHNpXmEMbboxJ6ihFlXcmVoQEx3HwBqdfE6Fk5oRu744aDApbdUc0J4WK/Cv4xcjw2WRG1b7TUjHFzfQMbziuNlik/o65zDjEo0WKfJLKZ/P6CweWtSREMLwqbjkt+RbOBdFcu79FcwPef5jE6fpddipKxOWZq+QZ05Kesap/v7OPc/pKAAelbeIZa3gtJesOVzLHXcd3O1620hy3lVYCZjJgVroRVhOxO6S9w9Nylo2Tt1xVipwr3CCDNv97qp5KyFczrQ1UYlJu/iqTcP6YJtQvGZNDtS5ZHul+8JdDU4osrsYHE3Gkpt9jPD9Kxhcg5brdL3AgMBAAGjggM7MIIDNzAdBgNVHQ4EFgQU2ayFg7boLF++hn+cIDBP8AuBYL4wHwYDVR0jBBgwFoAUDIanuimnqw/y+njgpCNP4oCVxbYwggErBgNVHR8EggEiMIIBHjCCARqgggEWoIIBEoaBymxkYXA6Ly8vQ049Q2l0eU9mQm9zdG9uLUVudGVycHJpc2UtU3ViLUNBLENOPVpQQ09CU0NBQVBQMDEsQ049Q0RQLENOPVB1YmxpYyUyMEtleSUyMFNlcnZpY2VzLENOPVNlcnZpY2VzLENOPUNvbmZpZ3VyYXRpb24sREM9Ym9zdG9uLERDPWNvYj9jZXJ0aWZpY2F0ZVJldm9jYXRpb25MaXN0P2Jhc2U/b2JqZWN0Q2xhc3M9Y1JMRGlzdHJpYnV0aW9uUG9pbnSGQ2h0dHA6Ly9jcmwuYm9zdG9uLmNvYi9DZXJ0RW5yb2xsL0NpdHlPZkJvc3Rvbi1FbnRlcnByaXNlLVN1Yi1DQS5jcmwwggFEBggrBgEFBQcBAQSCATYwggEyMIG8BggrBgEFBQcwAoaBr2xkYXA6Ly8vQ049Q2l0eU9mQm9zdG9uLUVudGVycHJpc2UtU3ViLUNBLENOPUFJQSxDTj1QdWJsaWMlMjBLZXklMjBTZXJ2aWNlcyxDTj1TZXJ2aWNlcyxDTj1Db25maWd1cmF0aW9uLERDPWJvc3RvbixEQz1jb2I/Y0FDZXJ0aWZpY2F0ZT9iYXNlP29iamVjdENsYXNzPWNlcnRpZmljYXRpb25BdXRob3JpdHkwcQYIKwYBBQUHMAKGZWh0dHA6Ly9haWEuYm9zdG9uLmNvYi9DZXJ0RW5yb2xsL1pQQ09CU0NBQVBQMDEuY2l0eWhhbGwuYm9zdG9uLmNvYl9DaXR5T2ZCb3N0b24tRW50ZXJwcmlzZS1TdWItQ0EuY3J0MA4GA1UdDwEB/wQEAwIFoDA8BgkrBgEEAYI3FQcELzAtBiUrBgEEAYI3FQiG54BehtyeT4eZmyaF9OcShcyXUF+L62mH7PtvAgFkAgEKMBMGA1UdJQQMMAoGCCsGAQUFBwMBMBsGCSsGAQQBgjcVCgQOMAwwCgYIKwYBBQUHAwEwDQYJKoZIhvcNAQELBQADggIBAAwojfsctWBq3KeYXWzlMtIzB6c0xLOFvw/hcsKT+n7oQG1RbN4For/SJtbtdE7cswNjOc7ZKI9KiRzSJSBPtVmiRfyaCMeNLbkXqrYJ8/Ke/APOuVJhBtbm5WleNUovr0BscVC/5bOv8PNlmenS9DPjucvHoghX++91+od7zca2kzavvP5TZhmc0+ls//dFwjPrKQudb2Bo3y93z6sUh+zaadvMv3jwKMyLWY2oEmaUIjMRyuPMKKbOFb0U+rIfB49vsAc3HqVMFf5mx0O7fVot91V7zfp1hNGm7CXMyKuMh8MRuwHxPPL0GmRD0PRNs2KCWSd1fnkFM5u8gDIQyGO/j88Epc+zWMiWYyQVc+7qvCVuDZh2cWaeEiNeOrOiPbkrtftMf1IxO091IwqGKEfEo8cN+d4EDxMMR8hUcKnZ+wXoKwfzXaXX++7ARwBeEtVASqNw5r7ncY/3LEZPwDYyDPIMubjZtpZeSzL71G0AWKGXUcAU0TReZyLGGWTdojoiuIsVifaMZNpMVEFQ5+xWw49XmXIBGuZYyWLf1SO87baq8J8/w5EnMnJcbsB+oWtzUvu+td6UvRup48VRBabtr9FM3QtI4GkjwVoZWihHMYK92iThcZShm5mBSLvSDsoKSRYU/zOKy4ZQ473zGDDp3AsBC5tsnNmqYnaSDXIN
-----END CERTIFICATE-----",
  ],
  "shared_options": Object {
    "sign_get_request": false,
  },
  "sso_login_url": "https://sso-dev.boston.gov/idp/SSO.saml2",
  "sso_logout_url": "https://sso-dev.boston.gov/idp/SSO.saml2",
}
`;
